File: bench/hash/hash.c
#include "../_lib/bench.h"

#define N        16384
#define N_ITERS  700
#define N_RUNS   21
#define N_WARMUP 5

static const uint32_t C1 = 0xcc9e2d51u;
static const uint32_t C2 = 0x1b873593u;

static uint8_t  buf[N];
static double   samples[N_RUNS];

static void init_buf(void) {
  int32_t x = 0x12345678;
  for (int i = 0; i < N; i++) {
    x = (int32_t)((int32_t)(x * (int32_t)1103515245) + 12345);
    buf[i] = (uint8_t)(((uint32_t)x >> 16) & 0xffu);
  }
}

static uint32_t murmur3(const uint8_t* buf, int n, uint32_t seed) {
  int32_t h = (int32_t)seed;
  for (int i = 0; i + 4 <= n; i += 4) {
    int32_t k = (int32_t)((uint32_t)buf[i]
                        | ((uint32_t)buf[i+1] << 8)
                        | ((uint32_t)buf[i+2] << 16)
                        | ((uint32_t)buf[i+3] << 24));
    k = (int32_t)((uint32_t)k * C1);
    k = (int32_t)(((uint32_t)k << 15) | ((uint32_t)k >> 17));
    k = (int32_t)((uint32_t)k * C2);
    h ^= k;
    h = (int32_t)(((uint32_t)h << 13) | ((uint32_t)h >> 19));
    h = (int32_t)((uint32_t)(int32_t)((uint32_t)h * 5u) + 0xe6546b64u);
  }
  h ^= n;
  uint32_t uh = (uint32_t)h;
  uh ^= uh >> 16;
  uh = (uint32_t)(uh * 0x85ebca6bu);
  uh ^= uh >> 13;
  uh = (uint32_t)(uh * 0xc2b2ae35u);
  uh ^= uh >> 16;
  return uh;
}

static uint32_t run_kernel(void) {
  uint32_t h = 0;
  for (int it = 0; it < N_ITERS; it++) {
    uint32_t mr = murmur3(buf, N, 0x9747b28cu);
    h = mix_u32(h, (uint32_t)(int32_t)mr);
    int j = it % N;
    buf[j] = (uint8_t)((buf[j] + 1u) & 0xffu);
  }
  return h;
}

int main(void) {
  init_buf();
  uint32_t cs = 0;
  for (int i = 0; i < N_WARMUP; i++) cs = run_kernel();
  for (int i = 0; i < N_RUNS; i++) {
    double t0 = now_ms();
    cs = run_kernel();
    samples[i] = now_ms() - t0;
  }
  print_result(median_us(samples, N_RUNS), cs, N * N_ITERS, 1, N_RUNS);
}

File: bench/hash/hash.go
package main

import (
	"fmt"
	"time"
)

const (
	n       = 16384
	nIters  = 700
	nRuns   = 21
	nWarmup = 5
)

const c1 uint32 = 0xcc9e2d51
const c2 uint32 = 0x1b873593

func mix(h, x uint32) uint32 { return (h ^ x) * 0x01000193 }

func medianUs(samples []float64) int {
	for i := 1; i < len(samples); i++ {
		v := samples[i]
		j := i - 1
		for j >= 0 && samples[j] > v {
			samples[j+1] = samples[j]
			j--
		}
		samples[j+1] = v
	}
	return int(samples[(len(samples)-1)>>1] * 1000)
}

func initBuf(buf []byte) {
	x := int32(0x12345678)
	for i := range buf {
		x = x*1103515245 + 12345
		buf[i] = byte((uint32(x) >> 16) & 0xff)
	}
}

func murmur3(buf []byte, bLen int, seed uint32) uint32 {
	h := int32(seed)
	for i := 0; i+4 <= bLen; i += 4 {
		k := int32(uint32(buf[i]) | uint32(buf[i+1])<<8 | uint32(buf[i+2])<<16 | uint32(buf[i+3])<<24)
		k = int32(uint32(k) * c1)
		k = int32(uint32(k)<<15 | uint32(k)>>17)
		k = int32(uint32(k) * c2)
		h ^= k
		h = int32(uint32(h)<<13 | uint32(h)>>19)
		h = int32(uint32(h)*5 + 0xe6546b64)
	}
	h ^= int32(bLen)
	uh := uint32(h)
	uh ^= uh >> 16
	uh *= 0x85ebca6b
	uh ^= uh >> 13
	uh *= 0xc2b2ae35
	uh ^= uh >> 16
	return uh
}

func runKernel(buf []byte) uint32 {
	h := uint32(0)
	for it := 0; it < nIters; it++ {
		mr := murmur3(buf, n, 0x9747b28c)
		h = mix(h, mr)
		j := it % n
		buf[j] = (buf[j] + 1) & 0xff
	}
	return h
}

func main() {
	buf := make([]byte, n)
	initBuf(buf)
	cs := uint32(0)
	for i := 0; i < nWarmup; i++ {
		cs = runKernel(buf)
	}
	samples := make([]float64, nRuns)
	for i := 0; i < nRuns; i++ {
		t0 := time.Now()
		cs = runKernel(buf)
		samples[i] = float64(time.Since(t0).Nanoseconds()) / 1e6
	}
	fmt.Printf("median_us=%d checksum=%d samples=%d stages=%d runs=%d\n",
		medianUs(samples), cs, n*nIters, 1, nRuns)
}
